File: src/main/java/ru/Controllers/Regulator/RegulatorController.java
package ru.Controllers.Regulator;

import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.scene.control.ComboBox;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import ru.Controllers.MainController;
import ru.Controllers.Signal.SignalTypes;
import ru.Models.RegulatorModel;
import ru.Models.SignalModel;
import ru.Utils.Utils;

public class RegulatorController {
    private Label amplitudeLabel;
    private TextField amplitudeTextField;
    private Label dcLabel;
    private TextField dcTextField;
    private Label frequencyLabel;
    private Label noiseLabel;
    private ComboBox<String> noiseTypesComboBox;
    private TextField frequencyTextField;
    private Label phaseLabel;
    private TextField phaseTextField;
    private boolean regulatorEnabled;
    private TextField responseAmplitudeTextField;
    private TextField responseDcTextField;
    private TextField responseFrequencyTextField;
    private TextField responseRmsTextField;
    private Label signalSettingsLabel;
    private Label signalTypeLabel;
    private ComboBox<String> signalTypeComboBox;

    private RegulatorModel regulatorModel = new RegulatorModel();
    private MainController mainController;

    public void initialize(MainController mainController) {
        amplitudeLabel = mainController.getAmplitudeLabel();
        amplitudeTextField = mainController.getAmplitudeTextField();
        dcLabel = mainController.getDcLabel();
        dcTextField = mainController.getDcTextField();
        frequencyLabel = mainController.getFrequencyLabel();
        frequencyTextField = mainController.getFrequencyTextField();
        noiseLabel = mainController.getNoiseLabel();
        noiseTypesComboBox = mainController.getNoiseTypesComboBox();
        phaseLabel = mainController.getPhaseLabel();
        phaseTextField = mainController.getPhaseTextField();
        responseAmplitudeTextField = mainController.getReceivedAmplitudeTextField();
        responseDcTextField = mainController.getReceivedDCTextField();
        responseFrequencyTextField = mainController.getReceivedFrequencyTextField();
        responseRmsTextField = mainController.getReceivedRmsTextField();
        signalSettingsLabel = mainController.getSignalSettingsLabel();
        signalTypeLabel = mainController.getSignalTypeLabel();
        signalTypeComboBox = mainController.getSignalTypeComboBox();
        this.mainController = mainController;
    }

    public void toggleRegulator(boolean isEnable) {
        regulatorEnabled = isEnable;
        listenTextFields();
        listenNeededParameterComboBox();
        toggleNeededParameter();

        if (isEnable) {
            Platform.runLater(() -> {
                signalSettingsLabel.setText("Настройки регулятора:");
                amplitudeLabel.setText("П - коэффициент:");
                amplitudeTextField.setText(String.valueOf(regulatorModel.getpCoefficient()));
                frequencyLabel.setText("И - коэффициент:");
                frequencyTextField.setText(String.valueOf(regulatorModel.getiCoefficient()));
                phaseLabel.setText("Д - коэффициент:");
                phaseTextField.setText(String.valueOf(regulatorModel.getdCoefficient()));
            });
        } else {
            Platform.runLater(() -> {
                signalSettingsLabel.setText("Настройки генерируемого сигнала:");
                amplitudeLabel.setText("Амплитуда, В:");
                amplitudeTextField.setText(String.valueOf(mainController.getSignalController().getSignalModel().getAmplitude()));
                frequencyLabel.setText("Частота, Гц:");
                frequencyTextField.setText(String.valueOf(mainController.getSignalController().getSignalModel().getFrequency()));
                phaseLabel.setText("Фаза, °:");
                phaseTextField.setText(String.valueOf(mainController.getSignalController().getSignalModel().getPhase()));
            });
        }

        Platform.runLater(() -> {
            amplitudeLabel.setDisable(!isEnable);
            amplitudeTextField.setDisable(!isEnable);
            frequencyLabel.setDisable(!isEnable);
            frequencyTextField.setDisable(!isEnable);
            phaseLabel.setDisable(!isEnable);
            phaseTextField.setDisable(!isEnable);
            noiseLabel.setVisible(!isEnable);
            noiseTypesComboBox.setVisible(!isEnable);
        });
    }

    private void listenTextFields() {
        if (regulatorEnabled) {
            amplitudeTextField.textProperty().addListener((observable, oldValue, newValue) -> {
                if (regulatorEnabled) {
                    regulatorModel.setpCoefficient(Double.parseDouble(newValue));
                }
            });

            frequencyTextField.textProperty().addListener((observable, oldValue, newValue) -> {
                if (regulatorEnabled) {
                    regulatorModel.setiCoefficient(Double.parseDouble(newValue));
                }
            });

            phaseTextField.textProperty().addListener((observable, oldValue, newValue) -> {
                if (regulatorEnabled) {
                    regulatorModel.setdCoefficient(Double.parseDouble(newValue));
                }
            });

            dcTextField.textProperty().addListener((observable, oldValue, newValue) -> {
                if (regulatorEnabled) {
                    regulatorModel.setNeededParameter(Double.parseDouble(newValue));
                }
            });
        }
    }

    private void listenNeededParameterComboBox() {
        signalTypeComboBox.valueProperty().addListener(observable -> {
            if (regulatorEnabled) {
                regulatorModel.setSelectedParameter(signalTypeComboBox.getSelectionModel().getSelectedItem());
            }
        });
    }

    private void toggleNeededParameter() {
        if (regulatorEnabled) {
            String selectedParameter = regulatorModel.getSelectedParameter();

            Platform.runLater(() -> {
                dcLabel.setText("Норма:");
                dcTextField.setText(String.valueOf(regulatorModel.getNeededParameter()));
                signalTypeLabel.setText("Регулировать по:");
                setNeededParameters();
                signalTypeComboBox.getSelectionModel().select(selectedParameter);
            });
        } else {
            String selectedType = mainController.getSignalController().getSignalType();

            Platform.runLater(() -> {
                dcLabel.setText("Статика, В:");
                dcTextField.setText(String.valueOf(mainController.getSignalController().getSignalModel().getDc()));
                signalTypeLabel.setText("Тип сигнала:");
                setSignalTypes();
                signalTypeComboBox.getSelectionModel().select(selectedType);
            });
        }

        Platform.runLater(() -> {
            dcLabel.setDisable(!regulatorEnabled);
            dcTextField.setDisable(!regulatorEnabled);
            signalTypeLabel.setDisable(!regulatorEnabled);
            signalTypeComboBox.setDisable(!regulatorEnabled);
        });
    }

    private void setNeededParameters() {
        ObservableList<String> parameters = FXCollections.observableArrayList();

        parameters.add(NeededParameters.AMPLITUDE.getParameterName());
        parameters.add(NeededParameters.DC.getParameterName());
        parameters.add(NeededParameters.FREQUENCY.getParameterName());
        parameters.add(NeededParameters.RMS.getParameterName());

        signalTypeComboBox.setItems(parameters);
    }

    private void setSignalTypes() {
        ObservableList<String> types = FXCollections.observableArrayList();

        types.add(SignalTypes.SINE.getTypeName());
        types.add(SignalTypes.PULSE.getTypeName());
        types.add(SignalTypes.TRIANGLES.getTypeName());
        types.add(SignalTypes.SAW.getTypeName());
        types.add(SignalTypes.NOISE.getTypeName());

        mainController.getSignalTypeComboBox().setItems(types);
    }


    public RegulatorModel getRegulatorModel() {
        return regulatorModel;
    }
}

File: src/main/java/ru/controllers/regulator/RegulatorController.kt
package ru.controllers.regulator

import javafx.application.Platform
import javafx.collections.FXCollections
import javafx.scene.control.ComboBox
import javafx.scene.control.TextField
import ru.controllers.MainController
import ru.controllers.signal.SignalTypes
import ru.models.RegulatorModel
import ru.utils.Utils


class RegulatorController(val mainController: MainController) {
    var isRegulatorEnabled = false
    private val regulatorModel = RegulatorModel()

    fun toggleRegulator(isEnable: Boolean) {
        isRegulatorEnabled = isEnable
        listenTextFields()
        listen(mainController.signalTypesComboBox)
        toggleUiElements()
        correctControlledParameter()
    }

    private fun listenTextFields() {
        listen(mainController.amplitudeTextField) { value -> regulatorModel.pCoefficient = value }
        listen(mainController.frequencyTextField) { value -> regulatorModel.iCoefficient = value }
        listen(mainController.phaseTextField) { value -> regulatorModel.dCoefficient = value }
        listen(mainController.dcTextField) { value -> regulatorModel.setNeededValueOfControlledParameter(value) }
    }

    private fun listen(textField: TextField, option: (Double) -> Unit) {
        textField.textProperty().addListener { _, _, newValue ->
            if (isRegulatorEnabled && textField.text.isNotEmpty() && textField.text != "-") {
                option(newValue.toDouble())
            }
        }
    }

    private fun listen(comboBox: ComboBox<String>) {
        comboBox.valueProperty().addListener { _ ->
            if (isRegulatorEnabled) {
                when (comboBox.selectionModel.selectedItem) {
                    ControlledParameters.AMPLITUDE.parameterName -> regulatorModel.selectedParameter = ControlledParameters.AMPLITUDE
                    ControlledParameters.DC.parameterName -> regulatorModel.selectedParameter = ControlledParameters.DC
                    ControlledParameters.FREQUENCY.parameterName -> regulatorModel.selectedParameter = ControlledParameters.FREQUENCY
                }
                regulatorModel.setNeededValueOfControlledParameter(parseNeededValueOfControlledParameter(mainController.dcTextField))
            }
        }
    }

    private fun parseNeededValueOfControlledParameter(textField: TextField): Double {
        return if (textField.text.isNotEmpty() && textField.text != "-") textField.text.toDouble() else 0.0
    }

    private fun toggleUiElements() {
        if (isRegulatorEnabled) {
            Platform.runLater {
                mainController.signalSettingsLabel.text = "Настройки регулятора:"
                mainController.amplitudeLabel.text = "П - коэффициент:"
                mainController.amplitudeTextField.text = regulatorModel.pCoefficient.toString()
                mainController.frequencyLabel.text = "И - коэффициент:"
                mainController.frequencyTextField.text = regulatorModel.iCoefficient.toString()
                mainController.phaseLabel.text = "Д - коэффициент:"
                mainController.phaseTextField.text = regulatorModel.dCoefficient.toString()
                mainController.dcLabel.text = "Норма:"
                mainController.dcTextField.text = regulatorModel.getNeededValueOfControlledParameter().toString()
                mainController.signalTypeLabel.text = "Регулировать по:"
                setControlledParameters()
            }
        } else {
            Platform.runLater {
                mainController.signalSettingsLabel.text = "Настройки генерируемого сигнала:"
                mainController.amplitudeLabel.text = "Амплитуда, В:"
                mainController.amplitudeTextField.text = mainController.signalController.getAmplitude().toString()
                mainController.frequencyLabel.text = "Частота, Гц:"
                mainController.frequencyTextField.text = mainController.signalController.getFrequency().toString()
                mainController.phaseLabel.text = "Фаза, °:"
                mainController.phaseTextField.text = mainController.signalController.getPhase().toString()
                mainController.dcLabel.text = "Статика, В:"
                mainController.dcTextField.text = mainController.signalController.getDc().toString()
                mainController.signalTypeLabel.text = "Тип сигнала:"
                setSignalTypes()
            }
        }

        Platform.runLater {
            mainController.amplitudeLabel.isDisable = !isRegulatorEnabled
            mainController.amplitudeTextField.isDisable = !isRegulatorEnabled
            mainController.frequencyLabel.isDisable = !isRegulatorEnabled
            mainController.frequencyTextField.isDisable = !isRegulatorEnabled
            mainController.phaseLabel.isDisable = !isRegulatorEnabled
            mainController.phaseTextField.isDisable = !isRegulatorEnabled
            mainController.noiseLabel.isDisable = !isRegulatorEnabled
            mainController.noiseTypesComboBox.isDisable = !isRegulatorEnabled
            mainController.dcLabel.isDisable = !isRegulatorEnabled
            mainController.dcTextField.isDisable = !isRegulatorEnabled
            mainController.signalTypeLabel.isDisable = !isRegulatorEnabled
            mainController.signalTypesComboBox.isDisable = !isRegulatorEnabled
        }
    }

    private fun setControlledParameters() {
        val parameters = FXCollections.observableArrayList<String>(ControlledParameters.AMPLITUDE.parameterName,
                ControlledParameters.DC.parameterName, ControlledParameters.FREQUENCY.parameterName)
        mainController.signalTypesComboBox.items = parameters

        when (regulatorModel.selectedParameter) {
            ControlledParameters.AMPLITUDE -> mainController.signalTypesComboBox.selectionModel.select(ControlledParameters.AMPLITUDE.parameterName)
            ControlledParameters.DC -> mainController.signalTypesComboBox.selectionModel.select(ControlledParameters.DC.parameterName)
            ControlledParameters.FREQUENCY -> mainController.signalTypesComboBox.selectionModel.select(ControlledParameters.FREQUENCY.parameterName)
        }
    }

    private fun correctControlledParameter() {
        if (isRegulatorEnabled) {
            Thread {
                while (isRegulatorEnabled && !mainController.controllerManager.isFinished) {
                    when (regulatorModel.selectedParameter) {
                        ControlledParameters.AMPLITUDE -> {
                            val oldValue = mainController.signalController.getAmplitude()
                            mainController.signalController.setAmplitude(oldValue + regulatorModel.getCorrection())
                        }
                        ControlledParameters.DC -> {
                            val oldValue = mainController.signalController.getDc()
                            mainController.signalController.setDc(oldValue + regulatorModel.getCorrection())
                        }
                        ControlledParameters.FREQUENCY -> {
                            val oldValue = mainController.signalController.getFrequency()
                            mainController.signalController.setFrequency(oldValue + regulatorModel.getCorrection())
                        }
                    }
                    Utils.sleep(1000)
                }
            }.start()
        }
    }

    private fun setSignalTypes() {
        val types = FXCollections.observableArrayList<String>(SignalTypes.SINE.typeName,
                SignalTypes.PULSE.typeName, SignalTypes.TRIANGLES.typeName, SignalTypes.SAW.typeName,
                SignalTypes.NOISE.typeName)
        mainController.signalTypesComboBox.items = types
        mainController.signalTypesComboBox.selectionModel.select(mainController.signalController.signalType.typeName)
    }

    fun setResponseAmplitude(value: Double) {
        regulatorModel.responseAmplitude = value
    }

    fun setResponseDc(value: Double) {
        regulatorModel.responseDc = value
    }

    fun setResponseFrequency(value: Double) {
        regulatorModel.responseFrequency = value
    }
}